### ISSUE TITLE
Board definitions - external location?

### ISSUE THREAD

**Issue by 0xc0170** [0xc0170/project_generator#67]
In the #52, it was proposed by @aethaniel , to have board definitions outside of the project_generator. Or at least to be able to pass the yaml record with board definitions for a project, then use that information internally to get the right information.

```
    board_def = {
        # Freescale FRDM boards
        'frdm-k20d50m': {
            'uvision': 'MK20DX128xxx5',
            'gcc' : 'cortex-m4',
        },
        'frdm-kl25z' : {
            'iar' : 'MKL25Z128xxx4',
            'gcc' : 'cortex-m0+',
        },
        'frdm-k64f' : {
            'uvision' : 'MK64FN1M0xxx12',
            'gcc' : 'cortex-m4f',
        },

        # mbed boards
        'mbed-lpc1768' : {
            'uvision' : 'LPC1768',
            'iar' : 'LPC1768',
            'coide' : 'LPC1768',
            'gcc' : 'cortex-m3',
        },

        # ST Disco boards
        'disco_f407vg' : {
            'coide' : 'STM32F407VG',
            'gcc' : 'cortex-m4',
        },
    }
``` 

This can be expressed in the yaml file, provided externally, and the pg just gets the board. This ticket is to start a discussion how it can be implemented so it scales. The initial idea was to have a common place, but probably would be better to have a repo with all these definitions, which can be easily pulled and used.

**Comment by matthewelse** [comment]
I think this would make the most sense as a repo called project_generator_definitions or something like that

**Comment by aethaniel** [comment]
+1
Indeed, if it is possible it would be great and separate process from generic and global data.


On 18 March 2015 19:42:44 CET, Matthew Else <notifications@github.com> wrote:
>I think this would make the most sense as a repo called
>project_generator_definitions or something like that
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/0xc0170/project_generator/issues/67#issuecomment-83116994

-- 
Envoyé de mon téléphone Android avec K-9 Mail. Excusez la brièveté.

**Comment by 0xc0170** [comment]
I believe MCU def should be also then make external, the only tool settings definitions would stay.

The one idea can be a github repository, which can be automatically pulled and parsed , to form a valid python dictionary. Or supplied a path for find this definitions locally, and use those. Let's look at options, and propose a simpleyaml file how it would look like lets say for uvision and how to pass it ot the generator

**Comment by 0xc0170** [comment]
An example of yaml definitions


```
board_definitions:
    frdm-k20d50m:
        uvision:
            - MK20DX128xxx5
        gcc:
            - cortex-m4
    frdm-kl25z:
        iar:
            - MKL25Z128xxx4
        gcc:
            - cortex-m0+

mcu_definitions:
    uvision:
        MK20DX128xxx5:
            TargetOption:
                Device:
                    - MK20DX128xxx5
                Vendor:
                     'Freescale Semiconductor',
                Cpu:
                    - IRAM(0x1FFFE000-0x1FFFFFFF) IRAM2(0x20000000-0x20001FFF) IROM(0x0-0x1FFFF) IROM2(0x10000000-0x10007FFF) CLOCK(12000000) CPUTYPE("Cortex-M4") ELITTLE
                FlashDriverDll:
                    - ULP2CM3(-O2510 -S0 -C0 -FO15 -FD20000000 -FC800 -FN2 -FF0MK_P128_50MHZ -FS00 -FL020000 -FF1MK_D32_50MHZ -FS110000000 -FL108000)
                DeviceId:
                    - 6206
                SFDFile:
                    - SFD\Freescale\Kinetis\MK20D5.sfr
```

**Comment by matthewelse** [comment]
I think it would make more sense if the repo which contained all of the target definitions had for example:

> project_generator_devices
    |_FRDM_K64F.yaml
    |_LPC1768.yaml
    |_...

so each device (or class of devices) has its own yaml file... this could mean you could potentially have variants of devices which can be based on a different device

**Comment by aethaniel** [comment]
Be careful not mix devices and boards.

I would see a kind of database organized by silicon vendor with product families and according series.

As it is usual to have common points between series a base common file could be inherited/included by a specific series or part specific definitions file.

Then we would have the board referencing the device it is based on.

This is mainly what Matthew is proposing.

On top of that, applications have their own definitions like the memory mapping, but this is next story.

My two cents.

Thibaut

**Comment by 0xc0170** [comment]
The problem here is mcu name. I decided to follow tools name because each tool names MCU differently (many of MCU are the same but from 10 I have seen, 2 were different, like uvision names it XYZ and IAR XYX). I agree with boards, then we have to find how to name MCU.

```
project_generator_definitions/
    boards/
        frdm-k20d50m.yaml
        mbed-lpc1768.yaml
    mcu/
        uvision/
            MK20DX128xxx5.yaml
        iar/
            MK20DX128.yaml
        coide/
           K20D50M.yaml
```

This is not ideal.. I made up the example above, but there are cases when this is true. Each tool has own naming.. So ideally would be that mcu/mk20dx128xxx5 as that one seems to be accroding their reference manual. But who decides what names is correct? :)

**Comment by 0xc0170** [comment]
I am taking freescale as an example , as I had some problems with choosing the right chip in tools :-)

Based on comments above, is this the one we can start with?

```
project_generator_definitions/
    boards/
        frdm-k20d50m.yaml #this will need to describe: vendor: freescale mcu: mk20dx128xxx5
        mbed-lpc1768.yaml
    mcu/
        freescale/
            mk20dx128xxx5.yaml
        nxp/
            lpc1768.yaml
```

The one who adds a board, defines MCU.. So if I know that I have own board with the same mcu, and want to add different tool, I'll add it to the same file -> mk20dx128xxx5.yaml. If my MCU is not there yet, neither the board, I'll create a new one for both. A question is, what naming convection for mcu definition yaml file should I follow? The one is written in reference manual, one in my tool I use, or one I have written on my board?

**Comment by aethaniel** [comment]
Indeed, good point.

I would say the right name is the one provided by the vendor which is the one the user know from datasheet.
Then we can have an alias in device tool yaml.
Be careful that sometimes the name in tool can change, but not the one in DS.


On 20 March 2015 00:17:02 CET, Martin Kojtal <notifications@github.com> wrote:
>The problem here is mcu name. I decided to follow tools name because
>each tool names MCU differently (many of MCU are the same but from 10 I
>have seen, 2 were different, like uvision names it XYZ and IAR XYX). I
>agree with boards, then we have to find how to name MCU.
>
>```
>project_generator_definitions/
>    boards/
>        frdm-k20d50m.yaml
>        mbed-lpc1768.yaml
>    mcu/
>        uvision/
>            MK20DX128xxx5.yaml
>        iar/
>            MK20DX128.yaml
>        coide/
>           K20D50M.yaml
>```
>
>This is not ideal.. I made up the example above, but there are cases
>when this is true. Each tool has own naming.. So ideally would be that
>mcu/mk20dx128xxx5 as that one seems to be accroding their reference
>manual. But who decides what names is correct? :) 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/0xc0170/project_generator/issues/67#issuecomment-83799054

-- 
Envoyé de mon téléphone Android avec K-9 Mail. Excusez la brièveté.

**Comment by matthewelse** [comment]
I like the idea of the aliases in the yaml file, which can include the names for different boards in different toolchains - it sort of solves the problem @0xc0170 and I were talking about.

**Comment by aethaniel** [comment]
I see the repo already exists for definitions.
Can we begin to populate it a bit and then continue discussing there?

**Comment by 0xc0170** [comment]
That's what I am going to do today or during the weekend. I'll start porting what we have covered at the moment, boards like k64f, lpc1768 and others.

I added load command to the #70, so far no implementation, will be added once we agree on definitions structure, so I can write yaml parser for it.

**Comment by matthewelse** [comment]
These are the things I can think of that need doing

- [x] Implement load command
- [ ] Decide on format for device definitions
- [ ] Add board definitions to repository
- [ ] Implement using the downloaded definitions

**Comment by 0xc0170** [comment]
Almost all features from the above are implemented. One more to come,it should get a new command:

```
pgen update (default - to get it from official repository, or provide git remote address to get it from)
```

**Comment by 0xc0170** [comment]
Resolved via #77, look at https://github.com/0xc0170/project_generator_definitions
